examples/vertical-fl/README.md

```diff
@@ -0,0 +1,490 @@
+essentially feature embeddings that capture the patterns within each client's
+data slice. These embeddings are then ready to be sent to the server for further
+processing.
+
+#### Server
+
+The server's model acts as the central aggregator in the VFL system. It's also a
+neural network but with a slightly different architecture tailored to its role
+in aggregating the client models' outputs.
+
+```python3
+class ServerModel(nn.Module):
+    def __init__(self):
+        super(ServerModel, self).__init__()
+        self.fc = nn.Linear(12, 1)
+        self.sigmoid = nn.Sigmoid()
+
+    def forward(self, x):
+        x = self.fc(x)
+        return self.sigmoid(x)
+```
+
+It comprises a single linear layer that accepts the concatenated outputs from
+all client models as its input. The number of inputs to this layer equals the
+total number of outputs from the client models ( $3 x 4 = 12$ ). After processing
+these inputs, the linear layer's output is passed through a sigmoid activation
+function (`nn.Sigmoid()`), which maps the result to a `(0, 1)` range, providing
+a probability score indicative of the likelihood of survival.
+
+### Strategy
+
+The strategy we will write to perform the aggregation will inherit from `FedAvg`
+and set the following additional attributes:
+
+```python3
+self.model = ServerModel(12)
+self.initial_parameters = ndarrays_to_parameters(
+    [val.cpu().numpy() for _, val in self.model.state_dict().items()]
+)
+self.optimizer = optim.SGD(self.model.parameters(), lr=0.01)
+self.criterion = nn.BCELoss()
+self.label = torch.tensor(labels).float().unsqueeze(1)
+```
+
+With `labels` given as an argument to the strategy.
+
+We then redefine the `aggregate_fit` method:
+
+```python3
+def aggregate_fit(
+    self,
+    rnd,
+    results,
+    failures,
+):
+    # Do not aggregate if there are failures and failures are not accepted
+    if not self.accept_failures and failures:
+        return None, {}
+
+    # Convert results
+    embedding_results = [
+        torch.from_numpy(parameters_to_ndarrays(fit_res.parameters)[0])
+        for _, fit_res in results
+    ]
+    embeddings_aggregated = torch.cat(embedding_results, dim=1)
+    embedding_server = embeddings_aggregated.detach().requires_grad_()
+    output = self.model(embedding_server)
+    loss = self.criterion(output, self.label)
+    loss.backward()
+
+    self.optimizer.step()
+    self.optimizer.zero_grad()
+
+    grads = embedding_server.grad.split([4, 4, 4], dim=1)
+    np_grads = [grad.numpy() for grad in grads]
+    parameters_aggregated = ndarrays_to_parameters(np_grads)
+
+    with torch.no_grad():
+        correct = 0
+        output = self.model(embedding_server)
+        predicted = (output > 0.5).float()
+
+        correct += (predicted == self.label).sum().item()
+
+        accuracy = correct / len(self.label) * 100
+
+    metrics_aggregated = {"accuracy": accuracy}
+
+    return parameters_aggregated, metrics_aggregated
```

Comment:
Here is what I added to the README:
```
Note that this `aggregate_fit` function returns gradients instead of trained
weights. This is because, in this setting, sharing gradients allows each
participant to benefit from the collective feedback gathered from the entire
pool of data without the need to align their different feature spaces (trained
weights are directly tied to specific features of the dataset but not gradients,
which are just a measure of the sensitivity of the loss function to changes in
the model's parameters). This shared feedback, encapsulated in the gradients,
guides each participant's model to adjust and improve, achieving optimization
not just based on its own data but also leveraging insights from the entire
network's data.
```

What do you think?

Reply:
That's a great explanation! A small concern was the inconsistency of what we do here and what we said in the docstring. FYI, here's part of the docstring for `aggregate_fit`:

```
        Returns
        -------
        parameters : Tuple[Optional[Parameters], Dict[str, Scalar]]
            If parameters are returned, then the server will treat these as the
            new global model parameters (i.e., it will replace the previous
            parameters with the ones returned from this method). If `None` is
            returned (e.g., because there were only failures and no viable
            results) then the server will no update the previous model
            parameters, the updates received in this round are discarded, and
            the global model parameters remain the same.
```

We said in the docstring that the return value of `aggregate_fit`, if not None, is the new global model parameters. I think we should mention that we do not need to return parameters here is because updates are completed locally in VFL. Meanwhile, the server should send the gradients back to all clients to let them continue the back prop and update their local model. In flower, the parameters returned by `aggregate_fit` will be stored and sent to `Client.evaluate` via `configure_fit`. Thus, we can return gradients in our `aggregate_fit` and the returned gradients will be sent to `Client.evaluate` as `parameters`. That's also why we can obtain gradients from the `parameters` argument in `Client.evaluate`. 

I am not sure if I make myself clear, but the thing that I am concerned about is the mismatch between what is said in the docstring of `aggregate_fit` and how we actually use `aggregate_fit` here.

Reply:
Oh, yes I see your point! Thanks, I'll update the README

Reply:
Looks great!

Reply:
Thanks for the suggestions!